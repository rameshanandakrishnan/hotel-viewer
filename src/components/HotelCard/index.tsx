import React from 'react';

import Card from '@material-ui/core/Card';
import Rating from '@material-ui/lab/Rating';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';

import { HotelWithRooms } from '../../services/hotel';
import { makeStyles } from '@material-ui/core';

const useStyles = makeStyles(() => ({
  images: {
    width: '100%',
    height: '500px',
    objectFit: 'cover',
  },
  root: {
    width: '100%',
  },
}));

const HotelCard: React.FC<{ hotel: HotelWithRooms; renderAction: () => JSX.Element }> = ({
  hotel,
  renderAction,
}) => {
  const classes = useStyles();

  return (
    <Card className={classes.root} data-testref='hotel'>
      <CardMedia
        component='img'
        alt='Hotel Image'
        height='200'
        image={hotel.images[0].url}
        title='Hotel Image'
      />
      <CardContent>
        <Typography gutterBottom variant='h5' component='h2'>
          {hotel.name}
        </Typography>
        {Boolean(hotel.starRating) && <Rating value={parseInt(hotel.starRating)} readOnly={true} />}
        <Typography variant='body2' color='textSecondary' paragraph={true}>
          {hotel.town}/{hotel.country}
        </Typography>
        <Typography variant='body2' color='textSecondary'>
          Available Room Types for your search criteria : <strong>{hotel.rooms.length}</strong>
        </Typography>
      </CardContent>
      <CardActions>{renderAction()}</CardActions>
    </Card>
  );
};

export default HotelCard;
