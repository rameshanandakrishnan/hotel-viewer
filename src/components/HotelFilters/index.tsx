import React from 'react';

import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Toolbar from '@material-ui/core/Toolbar';
import { makeStyles, Typography } from '@material-ui/core';

import FilterListIcon from '@material-ui/icons/FilterList';
import Rating from '@material-ui/lab/Rating';
import StepIncrementer from '../StepIncrementor';

const MIN = 0;
const MAX = 10;

const useStyles = makeStyles(() => ({
  toolbar: {
    background: '#fff',
    borderBottom: '1px solid #E6EAEA',
    marginBottom: '16px',
    maxWidth: '100%',
    display: 'flex',
    flexWrap: 'nowrap',
    overflowX: 'auto',
  },
  separatorBox: {
    height: '65px',
    padding: '0 24px',
    borderRight: '1px solid #E6EAEA',
    display: 'flex',
    alignItems: 'center',
    flex: '0 0 auto',
  },
  label: {
    paddingRight: '8px',
  },
  resetButton: {
    marginLeft: '8px',
  },
}));

export type Filters = {
  ratings: number | null;
  adults: number;
  children: number;
};

type HotelFiltersProps = {
  onFilterChange: (filters: Filters) => void;
};

const filterDefault = {
  ratings: 0,
  adults: 0,
  children: 0,
};

const HotelFilters: React.FC<HotelFiltersProps> = ({ onFilterChange }) => {
  const [filters, setFilters] = React.useState<Filters>(filterDefault);
  const classes = useStyles();
  const didMountRef = React.useRef(false);

  const handleGuestChange = (
    { currentTarget: { name, value } }: React.MouseEvent<HTMLInputElement>,
    action: 'add' | 'remove'
  ) => {
    const guestType = name as 'children' | 'adults';

    setFilters((prevFilters) => ({
      ...prevFilters,
      [name]:
        action === 'add' ? (prevFilters[guestType] || 0) + 1 : (prevFilters[guestType] || 0) - 1,
    }));
  };

  const handleReset = () => {
    setFilters(filterDefault);
  };

  React.useEffect(() => {
    if (didMountRef.current) {
      onFilterChange(filters);
    } else {
      didMountRef.current = true;
    }
  }, [filters, onFilterChange]);

  return (
    <Toolbar className={classes.toolbar} id='hotelFilters'>
      <Box className={classes.separatorBox}>
        <FilterListIcon />
      </Box>

      <Box className={classes.separatorBox}>
        <Typography className={classes.label}>Star Rating: </Typography>
        <Rating
          name='rating'
          value={filters.ratings}
          onChange={(_, value) => {
            setFilters((prevFilters) => ({ ...prevFilters, ratings: value || 0 }));
          }}
          emptyLabelText='Select a Hotel Rating'
        />
      </Box>

      <Box className={classes.separatorBox}>
        <Typography className={classes.label}>Adults : </Typography>
        <StepIncrementer
          min={MIN}
          max={MAX}
          name='adults'
          value={filters.adults}
          onClick={handleGuestChange}
        />
      </Box>

      <Box className={classes.separatorBox}>
        <Typography className={classes.label}>Children : </Typography>
        <StepIncrementer
          min={MIN}
          max={MAX}
          name='children'
          value={filters.children}
          onClick={handleGuestChange}
        />
      </Box>

      <Button
        color='secondary'
        variant='outlined'
        onClick={handleReset}
        className={classes.resetButton}>
        Reset Filters
      </Button>
    </Toolbar>
  );
};

export default React.memo(HotelFilters);
