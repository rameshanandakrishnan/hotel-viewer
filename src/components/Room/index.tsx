import React from 'react';

import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core';
import Typography from '@material-ui/core/Typography';

import { Room } from '../../services/hotel';

const useStyles = makeStyles(() => ({
  images: {
    width: '100%',
    paddingTop: '8px',
  },
}));

const RoomDetails: React.FC<{ room: Room }> = ({ room }) => {
  const classes = useStyles();
  const adults = room.occupancy.maxAdults;
  const children = room.occupancy.maxOverall
    ? room.occupancy.maxOverall - room.occupancy.maxAdults
    : room.occupancy.maxChildren;
  const roomImage = room?.images?.length ? room.images[0].url : undefined;

  return (
    <Grid container={true} spacing={2} data-testref={`room-${room.id}`}>
      {!!roomImage && (
        <Grid item={true} md={4}>
          <img src={roomImage} alt='Room' className={classes.images} />
        </Grid>
      )}
      <Grid item={true} md={roomImage ? 8 : 12}>
        <Typography variant='h5'>{room.name}</Typography>
        <Typography variant='body1' paragraph={true}>
          Sleeps <strong>{adults}</strong> adults and <strong>{children}</strong>{' '}
          {children === 1 ? 'child' : 'children'}{' '}
          <small>
            <i>
              {room.occupancy.maxOverall ? ` ( Max Capacity: ${room.occupancy.maxOverall} )` : null}
            </i>
          </small>
        </Typography>
        <Typography paragraph={true} variant='body2'>
          {room.longDescription}
        </Typography>
      </Grid>
    </Grid>
  );
};

export default RoomDetails;
