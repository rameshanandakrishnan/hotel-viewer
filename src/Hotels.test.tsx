import React from 'react';
import fetchMock from 'fetch-mock';
import { mount } from 'enzyme';

import Hotel from './Hotels';
import HotelCard from './components/HotelCard';
import HotelFilters from './components/HotelFilters';
import Room from './components/Room';
import { Button, Dialog } from '@material-ui/core';

export const delay = (ms = 0) => new Promise((res) => setTimeout(res, ms));

describe('Hotel Page', () => {
  beforeAll(() => {
    fetchMock
      .get(`https://obmng.dbm.guestline.net/api/hotels?collection-id=OBMNG`, [
        {
          id: 'OBMNG1',
          name: 'Hotel 1',
          images: [{ url: 'http://pic.com/linkToOBMNG1Img.png' }],
          town: 'London',
          country: 'UK',
          starRating: '5',
        },
        {
          id: 'OBMNG2',
          name: 'Hotel 2',
          images: [{ url: 'http://pic.com/linkToOBMNG2Img.png' }],
          town: 'London',
          country: 'UK',
          starRating: '2',
        },
      ])
      .get('https://obmng.dbm.guestline.net/api/roomRates/OBMNG/OBMNG1', {
        rooms: [
          {
            id: 'ROOM_ID1',
            occupancy: { maxAdults: 3, maxChildren: 2, maxOverall: 4 },
            images: [{ url: 'http://pic.com/ROOM_ID1.png' }],
            longDescription: 'Long description HOTEL1 Room1',
          },
          {
            id: 'ROOM_ID2',
            occupancy: { maxAdults: 4, maxChildren: 1, maxOverall: 4 },
            longDescription: 'Long description HOTEL1 Room2',
          },
        ],
      })
      .get('https://obmng.dbm.guestline.net/api/roomRates/OBMNG/OBMNG2', {
        rooms: [
          {
            id: 'ROOM_ID1',
            occupancy: { maxAdults: 3, maxChildren: 2, maxOverall: 4 },
            images: [{ url: 'http://pic.com/ROOM_ID1.png' }],
            longDescription: 'Long description HOTEL2 Room 1',
          },
          {
            id: 'ROOM_ID2',
            occupancy: { maxAdults: 8, maxChildren: 5, maxOverall: 10 },
            longDescription: 'Long description HOTEL2 Room 2',
          },
        ],
      });
  });

  it('should render Hotels', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();

    expect(component).toMatchSnapshot();
  });

  it('should filter hotels by star rating', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 5, adults: 0, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);

    component.find(HotelFilters).props().onFilterChange({ ratings: 2, adults: 0, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);

    component.find(HotelFilters).props().onFilterChange({ ratings: 3, adults: 0, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(0);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 0, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(2);
  });

  it('should filter hotels based on adults', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 2, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 8, children: 0 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);
    expect(component.find(HotelCard).at(0).text()).toContain('Hotel 2');
  });

  it('should filter hotels based on children', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 0, children: 1 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 0, children: 5 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);
    expect(component.find(HotelCard).at(0).text()).toContain('Hotel 2');
  });

  it('should filter hotels based on adults and children and should take into consideration the max capacity', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 1, children: 1 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 8, children: 5 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(0);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 5, children: 5 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);
  });

  it('should filter hotels based on occupancy and star ratings', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 0, adults: 1, children: 1 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(2);

    component.find(HotelFilters).props().onFilterChange({ ratings: 5, adults: 1, children: 1 });

    component.update();

    expect(component.find(HotelCard).length).toEqual(1);
    expect(component.find(HotelCard).at(0).text()).toContain('Hotel 1');
  });

  it('should render dialog with rooms when the user click on view rooms', async () => {
    const component = mount(<Hotel collectionId='OBMNG' />);

    await delay();
    component.update();

    expect(component.find(Dialog).props().open).toEqual(false);
    expect(component.find(Dialog).find(Room).length).toEqual(0);

    component.find(HotelCard).at(0).find(Button).simulate('click');

    component.update();

    expect(component.find(Dialog).props().open).toEqual(true);
    expect(component.find(Dialog).find(Room).length).toEqual(2);
    expect(component.find(Dialog).find(Room).at(0).text()).toContain('HOTEL1 Room1');
    expect(component.find(Dialog).find(Room).at(1).text()).toContain('HOTEL1 Room2');
  });
});
