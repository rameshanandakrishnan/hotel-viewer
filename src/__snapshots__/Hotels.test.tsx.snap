// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotel Page should render Hotels 1`] = `ReactWrapper {}`;
